### PR TITLE
Project-Auth /Sari-Anne Peltola

### PR DESCRIPTION
**Pull request by sarianne-peltola**
FE: https://tender-cori-4354e6.netlify.app/login
BE: https://project-auth-sari.herokuapp.com/

Problem with FE part, something seems to be a problem with the absolute/relative positioning and can't type now in the input field!!! :O

### CODE REVIEW

File: frontend/src/reusable/urls.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+const BASE_URL = 'http://localhost:8080'
```
Comment (Bellsibub) [inline frontend/src/reusable/urls.js:1]: ❗❗ This need to be changed cause it makes it impossible to use the FE with the BE.

```suggestion
const BASE_URL = 'https://tender-cori-4354e6.netlify.app'
```

File: frontend/src/components/Login.js [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+    }
+    :focus + label {
+      bottom: 20px;
+      opacity: 1;
+      color: indigo;
+      font-size: 12px;
+      z-index: 1;
+    }
+  }
+`;
```
Comment (Bellsibub) [inline frontend/src/components/Login.js:212]: Very nice custom input ❤️

File: frontend/src/components/Login.js [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+            <Heart size={24} />
+            <InputBox
+              id='name'
+              maxLength='20'
+              type='text'
+              value={name}
+              onChange={(e) => setName(e.target.value)}
+              placeholder='Name'
+            />
+            <Label htlFor='name'>Name</Label>
```
Comment (Bellsibub) [inline frontend/src/components/Login.js:47]: ❕❕ The correct usage of Label and for in react is: (I think this just may have been a spell error 😅)
```suggestion
            <Label htmlFor='name'>Name</Label>
```
This also applies to line `63`

File: frontend/src/components/Main.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      const options = {
+        method: 'GET',
+        headers: {
+          Authorization: accessToken,
+        },
+      };
+
+      fetch(API_URL('mypage'), options)
+        .then((res) => res.json())
+        .then((data) => console.log(data));
```
Comment (Bellsibub) [inline frontend/src/components/Main.js:28]: ❕❕ There should be no developer console logs in production. Also not sure what the purpose of this fetch function is... Cause all it does is post the data in the console. Maybe you forgot to add something here?

File: frontend/src/reducers/user.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+      email: JSON.parse(localStorage.getItem('user')).email,
+      accessToken: JSON.parse(localStorage.getItem('user')).accessToken,
+      errors: null,
+    }
+  : {
+      name: null,
+      email: null,
+      accessToken: null,
+      errors: null,
+    };
```
Comment (Bellsibub) [inline frontend/src/reducers/user.js:18]: EPIC 🤯 😄

**Review by Bellsibub — CHANGES_REQUESTED** [review]
Great project! 😄 I really liked the design of the frontend and just the attention to detail 👍 

However there were some issues with connection of FE and BE. Therefore I couldn't properly test that part of the project.

Therefore I must request some changes on this. I've marked the comment related to this with ❗❗.

Also some minor issues I found in the code regarding console logs in production has also been marked with ❕❕. Since they are not application breaking I marked them with white 😉 
   
Hopefully my change suggestions helps you in resolving the issue with ease 😅

File: README.md [hunk 6]
```diff
@@ -1,13 +1,13 @@
...
 
 ## The problem
 
-Describe how you approached to problem, and what tools and techniques you used to solve it. How did you plan? What technologies did you use? If you had more time, what would be next?
+The styling is unfortunately not yet finished. I run into problems with my styled components and the absolute and relative positionings which I finally solved with z-index. The styling will be developed more, same with returning something meaningful from the api.
 
 ## View it live
 
-Every project should be deployed somewhere. Be sure to include the link to the deployed project so that the viewer can click around and see what it's all about.
+FE: https://tender-cori-4354e6.netlify.app/
```
Comment (KarolCislowski) [inline README.md:11]: To fix it you need to create `_redirects` file in the `public` directory of your react project and paste it inside:
``` 
/* /index.html 200
```


image

File: backend/server.js [hunk 7]
```diff
@@ -1,30 +1,121 @@
...
+  },
+  password: {
+    type: String,
+    required: true,
+  },
+  accessToken: {
+    type: String,
+    default: () => crypto.randomBytes(128).toString('hex'),
+  },
+});
```
Comment (KarolCislowski) [inline backend/server.js:38]: nice model

File: frontend/src/components/Header.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+  const accessToken = useSelector((store) => store.user.accessToken);
+  const dispatch = useDispatch();
+
+  const onButtonClick = () => {
+    batch(() => {
+      dispatch(user.actions.setAccessToken(null));
+      dispatch(user.actions.setName(null));
+      dispatch(user.actions.setEmail(null));
+
+      localStorage.removeItem('user')
```
Comment (KarolCislowski) [inline frontend/src/components/Header.js:16]: nice that you remembered to use localStorage

File: frontend/src/components/Main.js (hunk 4 above)
Comment (KarolCislowski) [reply frontend/src/components/Main.js:28]: 🧹 console.log 🧹

File: frontend/src/components/Main.js [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    }
+  }, [accessToken]);
```
Comment (KarolCislowski) [inline frontend/src/components/Main.js:30]: why 2 effects when you can have 1?

```jsx

useEffect(() => {
    if (!accessToken) {
      history.push('/login')
    } else {
      const options = { ...}
      ...
    }
  }, [accessToken, history, dispatch])

```

File: frontend/src/reducers/user.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+            );
+          });
+        } else {
+          dispatch(user.actions.setErrors(data));
+        }
+      });
+  };
+};
+
+export default user;
```
Comment (KarolCislowski) [inline frontend/src/reducers/user.js:75]: Powerful and easy-to-read reducer!

File: frontend/src/reusable/urls.js (hunk 1 above)
Comment (KarolCislowski) [inline frontend/src/reusable/urls.js:1]: this one line is a source of most problems in your app.

**Review by KarolCislowski — CHANGES_REQUESTED** [review]
Code looks great but you need to connect your frontend to the hosted backend and fix Netlify issue.

**Comment by KarolCislowski** [comment]
Because your graduation is so soon please let me know on slack (a short note with a link to this PR) when you finish fixing your app.

File: frontend/src/components/Main.js (hunk 4 above)
Comment (sarianne-peltola) [reply frontend/src/components/Main.js:28]: Yes I didn't have time to give it a reasonable functionality but wanted to leave it for further development.

**Comment by sarianne-peltola** [comment]
Requested changed done! Thank you soooo much!!!

**Review by KarolCislowski — APPROVED** [review]
Great APP 🥳

**Pull request closed without merging**